docs/merkletree/treetex/main.go

```diff
@@ -0,0 +1,235 @@
+// The caller may emit any number of child nodes before calling the returned
+// func to clode the node.
+// returns a func to be called to close the node.
+//
+func openInnerNode(prefix string, height, index, tier int64) func() {
+	attr := nInfo[nodeKey(height, index)].String()
+	fmt.Printf("%s [%d.%d, %s, tier=%d\n", prefix, height, index, attr, tier)
+	return func() { fmt.Printf("%s ]\n", prefix) }
+}
+
+// perfectInner renders the nodes of a perfect internal subtree.
+func perfectInner(prefix string, height, tier, index int64, top bool) {
+	nk := nodeKey(height, index)
+	setNodeInfo(nk, func(n *nodeInfo) { n.leaf = height == 0 })
+	setNodeInfo(nodeKey(height, index), func(n *nodeInfo) { n.perfectRoot = top })
+
+	if height == 0 {
+		drawLeaf(prefix, index)
+		return
+	}
+	c := openInnerNode(prefix, height, index, tier)
+	childIndex := index << 1
+	perfectInner(prefix+" ", height-1, tier-1, childIndex, false)
+	perfectInner(prefix+" ", height-1, tier-1, childIndex+1, false)
+	c()
+}
+
+// node renders a tree node.
+func node(prefix string, treeSize, height, tier, index int64) {
+	if height < 0 {
+		return
+	}
+	// Look at the bit of the treeSize corresponding to the current level:
+	b := (1 << uint(height)) & treeSize
+	rest := treeSize - b
+
+	if b > 0 {
+		// left child is a perfect subtree
+
+		// if there's a right-hand child, then we'll emit this node to be the
+		// parent. (Otherwise we'll just keep quiet, and recurse down - this is how
+		// we arrange for leaves to always be on the bottom level.)
+		if rest > 0 {
+			ch := height + 1
+			ci := index >> uint(ch)
+			setNodeInfo(nodeKey(ch, ci), func(n *nodeInfo) { n.ephemeral = true })
+			c := openInnerNode(prefix, ch, ci, tier)
+			defer c()
+		}
+		perfect(prefix+" ", height, tier-1, index>>uint(tier))
+		index += b
+	}
+	if rest == 0 {
+		return
+	}
+	node(prefix+" ", rest, height-1, tier-1, index)
+}
+
+// nodeKey returns a stable node identifier for the passed in node coordinate.
+func nodeKey(height, index int64) string {
+	return fmt.Sprintf("%d.%d", height, index)
+}
+
+// toNodeKey converts a storage.NodeID to the corresponding stable node
+// identifier used by this tool.
+func toNodeKey(n storage.NodeID) string {
+	d := int64(maxLen - n.PrefixLenBits)
+	i := n.BigInt().Int64() >> uint(d)
+	return nodeKey(d, i)
+}
+
+// Whee - here we go!
+func main() {
+	flag.Parse()
+	height := int64(bits.Len(uint(*treeSize-1)) + 1)
+
+	if *inclusion > 0 {
+		setNodeInfo(nodeKey(0, *inclusion), func(n *nodeInfo) { n.target = true })
+		nf, err := merkle.CalcInclusionProofNodeAddresses(*treeSize, *inclusion, *treeSize, maxLen)
+		if err != nil {
+			panic(err)
+		}
+		for _, n := range nf {
+			setNodeInfo(toNodeKey(n.NodeID), func(n *nodeInfo) { n.incProof = true })
+		}
+		h := int64(0)
+		i := *inclusion
+		for h < height {
+			setNodeInfo(nodeKey(h, i), func(n *nodeInfo) { n.incPath = true })
+			h++
+			i >>= 1
+		}
+	}
+
+	fmt.Print(preamble)
+	node("", *treeSize, height, height, 0)
```

Comment:
TODO Make the functions output to an `io` type?

Reply:
I'll add a TODO - I'd like to make it a `struct` first rather than pass another thing around